File: uptmdigital_app/lib/models/asignatura.dart
class Asignatura {
  final int idAsignatura;
  final String codigo;
  final String nombre;
  final int creditos;
  final int? semestreId;
  final String departamento;
  final int? profesorId;
  final int? carreraId;
  final String semestreNombre;

  Asignatura({
    required this.idAsignatura,
    required this.codigo,
    required this.nombre,
    required this.creditos,
    this.semestreId,
    required this.departamento,
    this.profesorId,
    this.carreraId,
    this.semestreNombre = '',
  });

  factory Asignatura.fromJson(Map<String, dynamic> json) {
    String sem = '';
    if (json['semestre'] is Map) {
      sem = json['semestre']['nombre']?.toString() ?? '';
    } else if (json['semestreNombre'] != null) {
      sem = json['semestreNombre'].toString();
    } else if (json['semestre'] != null) {
      sem = json['semestre'].toString();
    }

    return Asignatura(
      idAsignatura: json['idAsignatura'] ?? 0,
      codigo: json['codigo'] ?? '',
      nombre: json['nombre'] ?? '',
      creditos: json['creditos'] ?? 0,
      semestreId: json['semestreId'],
      departamento: json['departamento'] ?? '',
      profesorId: json['profesorId'],
      carreraId: json['carreraId'],
      semestreNombre: sem,
    );
  }
}

File: UPTMDigital.API/Models/Asignatura.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace UPTMDigital.API.Models
{
    [Table("Asignatura")]
    public class Asignatura
    {
        [Key]
        public int IdAsignatura { get; set; }
        public string Codigo { get; set; } = null!;
        public string Nombre { get; set; } = null!;
        public int Creditos { get; set; }
        public string? Departamento { get; set; }

        // --- FK Normalizadas ---
        public int? ProfesorId { get; set; }
        [ForeignKey("ProfesorId")]
        public Profesor? Profesor { get; set; }

        public int? SemestreId { get; set; }
        [ForeignKey("SemestreId")]
        public Semestre? Semestre { get; set; }

        public int? CarreraId { get; set; }
        [ForeignKey("CarreraId")]
        public Carrera? Carrera { get; set; }

        public bool Activo { get; set; } = true;
    }
}